storage: check leaseholder status in AdminSplit retry loop

nvb opened the pull request:
Fixes #23673.

This change adds a call to `redirectOnOrAcquireLease` to the AdminSplit
retry loop. This ensures that each attempt to perform a split is done
by a leaseholder. Without this check, it was possible that a stale
follower or a replica that had already been removed from the range
could end up in an infinite retry loop. This was possible because
these replicas could have stale knowledge about the state of the
range descriptor. The effect of this was that the replicas' updates
to the local range descriptor key could indefinitely hit
`ConditionFailedErrors`, which are retried.

The change also adds a regression test that creates this
"replica removed while splitting" scenario. Without the leaseholder
check introduced here, the test would fail.

Release note: None

nvb committed:
storage: add TestingRequestFilter testing knob

This change adds another testing filter that can be used to
influence the error returned from a request before it is evaluated.
Notably, the filter is run before the request is added to the
CommandQueue, so blocking in the filter will not block interfering
requests.

Release note: None

nvb committed:
storage: check leaseholder status in AdminSplit retry loop

Fixes #23673.

This change adds a call to `redirectOnOrAcquireLease` to the AdminSplit
retry loop. This ensures that each attempt to perform a split is done
by a leaseholder. Without this check, it was possible that a stale
follower or a replica that had already been removed from the range
could end up in an infinite retry loop. This was possible because
these replicas could have stale knowledge about the state of the
range descriptor. The effect of this was that the replicas' updates
to the local range descriptor key could indefinitely hit
`ConditionFailedErrors`, which are retried.

The change also adds a regression test that creates this
"replica removed while splitting" scenario. Without the leaseholder
check introduced here, the test would fail.

Now that the root cause of this infinite retry loop is fixed, we
should consider reverting #23310. This loop should not retry
indefinitely. If it does, exiting will just mask the issue.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/23762)
<!-- Reviewable:end -->

benesch commented:
> Now that the root cause of this infinite retry loop is fixed, we should consider reverting #23310. This loop should not retry indefinitely. If it does, exiting will just mask a more serious issue.

Well, the intent was to return the error up into the log file or jobs table where it would be visible to the user. Granted, I botched that 🙈, but I'm still nervous about retry loops that never bail out. Is this concern misplaced? This edge case went unnoticed for a while; I'm worried there might be another.

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

I think the limit on the retry loop should stay.

---

Reviewed 3 of 3 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica_command.go, line 716 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23762#-L7RK8ULYuVbcjwxl3dF:-L7RK8ULYuVbcjwxl3dG:b-8zgk86) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dda735ec2b1cc27cc5a5605dca193f34f302f73/pkg/storage/replica_command.go#L716)):*
> ```Go
> 	for retryable := retry.StartWithCtx(ctx, retryOpts); retryable.Next(); {
> 		// Admin commands always require the range lease to begin (see
> 		// executeAdminBatch), but we may have lost it while in this retry loop.
> ```

There are also entry points to AdminSplit that don't go through executeAdminBatch, IIRC. 

---

*[pkg/storage/store.go, line 669 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23762#-L7RJsOR8D6WSXbT2RD0:-L7RJsOR8D6WSXbT2RD1:b-b093oh) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dda735ec2b1cc27cc5a5605dca193f34f302f73/pkg/storage/store.go#L669)):*
> ```Go
> 	EvalKnobs batcheval.TestingKnobs
> 
> 	// TestingRequestFilter is called before evaluating each command.
> ```

Be more specific, since there's already EvalKnobs.TestingEvalFilter which runs at evaluation time. (the comment on ReplicaRequestFilter about when this runs belongs here, not on the function typedef)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23762#-:-L7RNPvdoArPBorz-UgH:b-kkbh3u)*
<!-- Sent from Reviewable.io -->

nvb committed:
storage: return an error from split retry loop after MaxRetries

This loop used to return a `nil` error once the retry loop was
exhausted, which made it look like the split had succeeded. It
now returns an error.

Release note: None

nvanbenschoten commented:
TFTRs! Added a new commit (now the second) to address https://github.com/cockroachdb/cockroach/pull/23310#issuecomment-372463203.

---

Review status: 2 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/storage/replica_command.go, line 716 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23762#-L7RK8ULYuVbcjwxl3dF:-L7UbSnzCutK72-Y8DIp:b-8pk9dt) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dda735ec2b1cc27cc5a5605dca193f34f302f73/pkg/storage/replica_command.go#L716)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

There are also entry points to AdminSplit that don't go through executeAdminBatch, IIRC. 
</blockquote></details>

It doesn't look like there are anymore. Everything else goes through the Replica by sending an `AdminSplitRequest`.

You might be thinking about the `splitQueue`, which calls `adminSplitWithDescriptor` directly. Because this bypasses `AdminSplit`, it is not subject to the retry loop. This looks like it was an intentional decision made in #10728 (see https://github.com/cockroachdb/cockroach/pull/10728#issuecomment-261653824).

---

*[pkg/storage/store.go, line 669 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23762#-L7RJsOR8D6WSXbT2RD0:-L7UdIOdMhq_7LkVaWST:bo3u0xo) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dda735ec2b1cc27cc5a5605dca193f34f302f73/pkg/storage/store.go#L669)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Be more specific, since there's already EvalKnobs.TestingEvalFilter which runs at evaluation time. (the comment on ReplicaRequestFilter about when this runs belongs here, not on the function typedef)
</blockquote></details>

Expanded this comment here but left the other because all of these function typedefs seem to have similar descriptions.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23762#-:-L7UkhtqRvb8ThiTebe3:bs1zafp)*
<!-- Sent from Reviewable.io -->
